test/Microsoft.ML.CpuMath.UnitTests.DotNetCoreApp/UnitTests.cs

```diff
@@ -0,0 +1,451 @@
+
+        [DllImport("CpuMathNative", EntryPoint = "SumSqU"), SuppressUnmanagedCodeSecurity]
+        internal unsafe static extern float NativeSumSqU(/*const*/ float* ps, int c);
+
+        [DllImport("CpuMathNative", EntryPoint = "AddU"), SuppressUnmanagedCodeSecurity]
+        internal unsafe static extern void NativeAddU(/*_In_ const*/ float* ps, /*_Inout_*/ float* pd, int c);
+
+        [DllImport("CpuMathNative", EntryPoint = "AddSU"), SuppressUnmanagedCodeSecurity]
+        internal unsafe static extern void NativeAddSU(/*_In_ const*/ float* ps, /*_In_ const*/ int* pi, /*_Inout_*/ float* pd, int c);
+
+        [DllImport("CpuMathNative", EntryPoint = "AddScaleU"), SuppressUnmanagedCodeSecurity]
+        internal unsafe static extern void NativeAddScaleU(float a, /*_In_ const*/ float* ps, /*_Inout_*/ float* pd, int c);
+
+        [DllImport("CpuMathNative", EntryPoint = "AddScaleSU"), SuppressUnmanagedCodeSecurity]
+        internal unsafe static extern void NativeAddScaleSU(float a, /*_In_ const*/ float* ps, /*_In_ const*/ int* pi, /*_Inout_*/ float* pd, int c);
+
+        [DllImport("CpuMathNative", EntryPoint = "ScaleU"), SuppressUnmanagedCodeSecurity]
+        internal unsafe static extern void NativeScaleU(float a, /*_Inout_*/ float* pd, int c);
+
+        [DllImport("CpuMathNative", EntryPoint = "Dist2"), SuppressUnmanagedCodeSecurity]
+        internal unsafe static extern float NativeDist2(/*const*/ float* px, /*const*/ float* py, int c);
+
+        [DllImport("CpuMathNative", EntryPoint = "SumAbsU"), SuppressUnmanagedCodeSecurity]
+        internal unsafe static extern float NativeSumAbsU(/*const*/ float* ps, int c);
+
+        [DllImport("CpuMathNative", EntryPoint = "MulElementWiseU"), SuppressUnmanagedCodeSecurity]
+        internal unsafe static extern void NativeMulElementWiseU(/*_In_ const*/ float* ps1, /*_In_ const*/ float* ps2, /*_Inout_*/ float* pd, int c);
+
+        [Theory]
+        [InlineData(10, 0, EXP_MAX)]
+        [InlineData(20, 0, EXP_MAX)]
+        [InlineData(20, 1, EXP_MAX / 2)]
+        [InlineData(30, 2, EXP_MAX / 2)]
+        public unsafe void DotUTest(int len, int seed, int expRange)
+        {
+            // arrange
+            float[] src = new float[len];
+            float[] dst = new float[len];
+            Random rand = new Random(seed);
+
+            for (int i = 0; i < len; i++)
+            {
+                src[i] = NextFloat(rand, expRange);
+                dst[i] = NextFloat(rand, expRange);
+            }
+
+            fixed (float* psrc = src)
+            fixed (float* pdst = dst)
+            {
+                // act
+                var nativeOutput = NativeDotU(psrc, pdst, len);
+                var managedOutput = CpuMathUtils.DotProductDense(src, dst, len);
+                output.WriteLine($"{nativeOutput} == {managedOutput}");
+
+                // assert
+                Assert.True(nativeOutput == managedOutput);
```

Comment:
I think this was a good approach for initially testing that the new methods produce the same output as the native methods.  However, I don't think this is a good long term approach. In the future, we some day hope to remove the native methods, if possible.  Also, as new methods get written to both the native and C# intrinsics, it would be quite possible that someone would write the same bug in both places.

Instead, I think the test should hard-code what it expects the results to be.

Reply:
Agree - that would save us from PInvoke calls in unit tests too.  Will fix that - thanks!